### PR TITLE
Get region information

### PR DESCRIPTION
**Pull request by bilalebi**
### JIRA Ticket
https://www.ebi.ac.uk/panda/jira/browse/EA-1094

### Changes
* Added gRPC endpoint which returns region information for given `genome_uuid`.
* And a 2nd endpoint which returns region information for given `genome_uuid` and `seq_region_name`

##### Get Genome Assembly Sequence

Request

```
{
    "genome_uuid": "2afef36f-3660-4b8c-819b-d1e5a77c9918",
    "chromosomal_only": false
}
```

Response

```
{
    {
        "name": "1",
        "rank": 1,
        "md5": "",
        "length": "216633511",
        "sha512t4u": "",
        "chromosomal": true
    },
    ...
    {
        "name": "LVXO01042241.1",
        "rank": 0,
        "md5": "",
        "length": "6941",
        "sha512t4u": "",
        "chromosomal": false
    },
    ...
}
```

##### Get Genome Assembly Sequence Region

Request

```
{
    "genome_uuid": "2afef36f-3660-4b8c-819b-d1e5a77c9918",
    "sequence_region_name": "3",
    "chromosomal_only": false
}
```

Response

```
{
    "name": "3",
    "md5": "",
    "length": "175453833",
    "sha512t4u": "",
    "chromosomal": true
}
```

### Related PR
* https://github.com/Ensembl/ensembl-metadata-service/pull/25
* https://github.com/Ensembl/ensembl-metadata-api/pull/31
* https://github.com/Ensembl/ensembl-metadata-api/pull/42

### CODE REVIEW

File: protos/ensembl/production/metadata/grpc/ensembl_metadata.proto [hunk 1]
```diff
@@ -216,6 +222,29 @@ message GenomeSequence {
...
 }
 
+/*
+Metadata about the sequences that comprise a genome's assembly.
+ */
+message GenomeAssemblySequence {
+  string name = 1;
+  int32 rank = 2;
+  string md5 = 3;
+  uint32 length = 4;
```
Comment (azangru) [inline protos/ensembl/production/metadata/grpc/ensembl_metadata.proto:232]: Max value of uint32 is ~4.29 billion. Is this going to be enough for whatever the longest chromosome is? There's a bit of a discussion [here](https://www.biostars.org/p/12560/), but might be safer to ask gene builders. Or switch to 64-bit integers.

Comment (bilalebi) [reply protos/ensembl/production/metadata/grpc/ensembl_metadata.proto:232]: I'll take a look, thank you for letting me know @azangru

**Comment by jgtate** [comment]
I'm probably missing something but why do we need to specify `assembly_accession` in the request? Isn't that implicit in the genome ID? Also, do we refer to `genome_uuid` anywhere else? Isn't it `genome_id` generally?

**Comment by bilalebi** [comment]
> I'm probably missing something but why do we need to specify `assembly_accession` in the request? Isn't that implicit in the genome ID? 

I think the expected inputs are not clear in my head, the ticked description says: 
> _a gRPC endpoint which for a given assembly will return a list of regions which are "top level"._

And

> for the given seq_region name in the given assembly.  If no region with that name exists, it should return a 404 or equivalent.

That's why I considered specifying `assembly_accession` (now I'm wondering why I added `genome_uuid`) 🤔  .. I'll need to talk to someone from web who will be using this endpoints

---
> Also, do we refer to `genome_uuid` anywhere else? Isn't it `genome_id` generally?

Sorry, I'm not sure I understand what you meant

**Comment by jgtate** [comment]
> That's why I considered specifying assembly_accession (now I'm wondering why I added genome_uuid) 🤔 .. I'll need to talk to someone from web who will be using these endpoints

Ultimately, the client is the thing that needs these data, with the metadata REST service acting as a proxy. Probably best to speak to @azangru about the inputs and outputs in the first instance.

> > Also, do we refer to genome_uuid anywhere else? Isn't it genome_id generally?

I just mean that we refer to the genome UUID when we're speaking about it, out loud in meetings, etc, but I would expect the code to call that piece of data `genome_id`, rather than `genome_uuid`. And for consistency it would be good to make sure that we use the same term across all of the APIs. 

Also, I know it's trivial, but calling it `uuid` would be leaking an implementation detail into the API, which I don't think we want to do.

**Comment by bilalebi** [comment]
> I just mean that we refer to the genome UUID when we're speaking about it, out loud in meetings, etc, but I would expect the code to call that piece of data `genome_id`, rather than `genome_uuid`. And for consistency it would be good to make sure that we use the same term across all of the APIs.
> 
> Also, I know it's trivial, but calling it `uuid` would be leaking an implementation detail into the API, which I don't think we want to do.

Thanks for bringing this to my attention, all endpoints are stating `genome_uuid`, which I don't mind changing to `genome_id` if necessary, I just need to let other teams know...

**Comment by azangru** [comment]
> Ultimately, the client is the thing that needs these data, with the metadata REST service acting as a proxy. Probably best to speak to @azangru about the inputs and outputs in the first instance.

Ideally, the client uses genome id to talk to the apis. Technically, we should have access to assembly accession id; but we would much rather use genome id everywhere.

Like John said, assembly accession id should be derivable from genome id.

Examples of current usage on the beta site:

- Client asks backend for a list of regions within karyotype:

https://beta.ensembl.org/api/genomesearch/genome/karyotype?genome_id=a7335667-93e7-11ec-a39d-005056b38ce3

- Client asks backend to validate user's input for region

https://beta.ensembl.org/api/genomesearch/genome/region/validate?genome_id=a7335667-93e7-11ec-a39d-005056b38ce3&region=13:32,352,475-32,352,533

File: protos/ensembl/production/metadata/grpc/ensembl_metadata.proto [hunk 2]
```diff
@@ -111,7 +111,10 @@ message AssemblyInfo {
   string level = 3;
   string name = 4;
   uint32 chromosomal = 5;
-  uint32 length = 6;
+  // length will be a string, because some browser can't handle a full 64bit value as a number,
+  // so the only way to ensure they can work for those clients is to handle them as strings
+  // source: https://github.com/protocolbuffers/protobuf/issues/2679
```
Comment (azangru) [inline protos/ensembl/production/metadata/grpc/ensembl_metadata.proto:116]: I'm confused... Where will it be a string? In a tool to inspect grpc responses, such as Insomnia? I think web metadata api can still send it as number. It seems that [javascript number type has 53 bit of precision](https://stackoverflow.com/a/858857/3925302)

Comment (bilalebi) [reply protos/ensembl/production/metadata/grpc/ensembl_metadata.proto:116]: > In a tool to inspect grpc responses, such as Insomnia?

Yes I added that comment after I noticed it was returning a string and tests were failing

**Comment by kamaldodiya** [comment]
Not getting any data for karyotyp yet. 
Tried following genome_uuid :     
# 9caa2cae-d1c8-4cfc-9ffd-2e13bc3e95b1 
# a733550b-93e7-11ec-a39d-005056b38ce3
# 2afef36f-3660-4b8c-819b-d1e5a77c9918

**Review by kamaldodiya — APPROVED** [review]

**Pull request merged**